lib/collection/src/operations/types.rs

```diff
@@ -584,7 +585,17 @@ impl From<RequestError<tonic::Status>> for CollectionError {
     fn from(err: RequestError<tonic::Status>) -> Self {
         match err {
             RequestError::FromClosure(status) => status.into(),
-            RequestError::Tonic(err) => CollectionError::service_error(format!("{err}")),
+            RequestError::Tonic(err) => {
+                let mut msg = err.to_string();
+                let mut err = err.source();
+
+                while let Some(src) = err {
+                    msg += &format!(": {src}");
+                    err = src.source();
+                }
```

Comment:
...and use it to build `msg` with fewer allocations.

```suggestion
                while let Some(src) = err {
                    err = src.source();
                    write!(&mut msg, ": {src}").unwrap(); // Writing into `String` never fails
                }
```

Reply:
Nice.

I did attempt a quick rewrite with the [`successors`](https://doc.rust-lang.org/std/iter/fn.successors.html) iterator as well, but lifetimes make it hard.

Reply:
> I did attempt a quick rewrite with the [successors](https://doc.rust-lang.org/std/iter/fn.successors.html) iterator as well, but lifetimes make it hard.

Awesome find! Works this way:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cf3463710648c5cd21b259f5e7470a04

Reply:
The types are a bit different, which makes it harder:

![image](https://github.com/qdrant/qdrant/assets/856222/f1e4312a-8858-400e-9ae8-35c20fd07623)

I wanted to use a `successors().map().collect().join(": ")` to make it a bit nicer. But whatever, Rust doesn't let me. :smiling_face_with_tear:

Reply:
Works like this then (notice `|&src|` in closure parameters):

```rust
for src in iter::successors(err.source(), |&src| src.source()) {
    write!(&mut msg, ": {src}").unwrap(); // Writing into `String` never fails
}
```

Reply:
Ugh, not sure how I missed the `&`. Thanks!

We can also do 😆:

```rust
let msg = std::iter::once(err.to_string())
    .chain(successors(err.source(), |&src| src.source()).map(ToString::to_string))
    .collect::<Vec<_>>()
    .join(": ");
CollectionError::service_error(msg)
```